node/src/components/linear_chain_sync.rs

```diff
@@ -520,8 +522,21 @@ where
                         // We shouldn't get invalid data from the storage.
                         // If we do, it's a bug.
                         assert_eq!(*block.hash(), block_hash, "Block hash mismatch.");
-                        trace!(%block_hash, "linear block found in the local storage.");
-                        self.block_downloaded(rng, effect_builder, &block)
+                        trace!(%block_hash, "Linear block found in the local storage.");
+                        // We hit a block that we already had in the storage - which should mean
+                        // that we also have all of its ancestors, so we switch to traversing the
+                        // chain forwards and downloading the deploys.
+                        // We don't want to download and execute a block we already have, so
+                        // instead of calling self.block_downloaded(), we take a shortcut:
+                        if let State::SyncingTrustedHash {
+                            ref mut latest_block,
+                            ..
+                        } = &mut self.state
+                        {
+                            *latest_block = Box::new(Some((&*block).clone()));
+                        }
+                        self.state.block_downloaded(&block);
```

Comment:
Can we extract this to a method and not modify `self`'s state directly in the component handlers?

Reply:
We could, but I wonder what that method should be called. `set_latest_block_if_syncing_trusted_hash`? :thinking: